### PR TITLE
support for multiple audiences

### PR DESCRIPTION
**Pull request by sunsided**
Added support for multiple valid audiences in both introspection and validation middleware. The audiences are implemented as `IList<string>` (i.e. ordered) so that a priority can be inferred for other uses (e.g. first is default).

### CODE REVIEW

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@
 using Microsoft.Extensions.Logging;
 using Microsoft.Net.Http.Headers;
 using Newtonsoft.Json.Linq;
+using System.Linq;
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: nit: `System` usings should go first :smile:

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs [hunk 2]
```diff
@@ -148,7 +149,7 @@ public class OAuthIntrospectionHandler : AuthenticationHandler<OAuthIntrospectio
         protected virtual Task<bool> ValidateAudienceAsync(JObject payload) {
             // If no explicit audience has been configured,
             // skip the default audience validation.
-            if (string.IsNullOrEmpty(Options.Audience)) {
+            if (Options.Audiences == null || Options.Audiences.Count <= 0) {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: Not sure I really like settable collection properties :smile: 

nit: `IList.Count` cannot be negative.

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs [hunk 3]
```diff
@@ -164,20 +165,20 @@ public class OAuthIntrospectionHandler : AuthenticationHandler<OAuthIntrospectio
                 case JTokenType.Array: {
                     // When the "aud" claim is an array, at least one value must correspond
                     // to the audience registered in the introspection middleware options.
-                    foreach (var audience in payload.Value<JArray>(OAuthIntrospectionConstants.Claims.Audience)) {
-                        if (string.Equals(audience.Value<string>(), Options.Audience, StringComparison.Ordinal)) {
-                            return Task.FromResult(true);
-                        }
+                    var audiences = payload.Value<JArray>(OAuthIntrospectionConstants.Claims.Audience).Select(audience => audience.Value<string>());
+                    if (audiences.Intersect(Options.Audiences, StringComparer.Ordinal).Any())
+                    {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: nit: brace style :smile:

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs (hunk 1 above)
Comment (sunsided) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: My bad

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionOptions.cs [hunk 4]
```diff
@@ -19,7 +20,7 @@ public class OAuthIntrospectionOptions : AuthenticationOptions {
         /// Setting this property is recommended when the authorization
         /// server issues access tokens for multiple distinct resource servers.
         /// </summary>
-        public string Audience { get; set; }
+        public IList<string> Audiences { get; set; }
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionOptions.cs]: Please consider removing the setter and assigning a default instance.
Updating the code documentation to indicate what happens when multiple audiences are set would be great :+1:

File: src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs [hunk 5]
```diff
@@ -60,7 +60,7 @@ public class OAuthValidationHandler : AuthenticationHandler<OAuthValidationOptio
         protected virtual Task<bool> ValidateAudienceAsync(AuthenticationTicket ticket) {
             // If no explicit audience has been configured,
             // skip the default audience validation.
-            if (string.IsNullOrEmpty(Options.Audience)) {
+            if (Options.Audiences == null || Options.Audiences.Count <= 0) {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/OAuthValidationHandler.cs]: Same remark.

File: src/AspNet.Security.OAuth.Validation/OAuthValidationOptions.cs [hunk 6]
```diff
@@ -17,7 +18,7 @@ public class OAuthValidationOptions : AuthenticationOptions {
         /// Setting this property is recommended when the authorization
         /// server issues access tokens for multiple distinct resource servers.
         /// </summary>
-        public string Audience { get; set; }
+        public IList<string> Audiences { get; set; }
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Validation/OAuthValidationOptions.cs]: Same remark.

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs (hunk 2 above)
Comment (sunsided) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: Well, `IList.Count` is `int`, so can be negative. :stuck_out_tongue_winking_eye:

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs (hunk 3 above)
Comment (sunsided) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: Hey, it's _your_ style that is mad! :smile: Will fix.

**Comment by PinpointTownes** [comment]
Great PR. Minor a few remarks, it's perfect :clap:

File: src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs (hunk 3 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs]: K&R FTW :trollface:

**Comment by PinpointTownes** [comment]
Nice. Squash it and I'll merge it :clap: 

Question: do you also want to submit a PR against the dev branch (for the OWIN version)?

**Comment by sunsided** [comment]
As for dev branch - can do :)

**Comment by PinpointTownes** [comment]
Merged, thanks Markus! :clap: 

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Extensions/commit/61115405a583acdae1654d3ec3cc70ccb5fb3ce3

**Comment by sunsided** [comment]
very welcome

**Pull request closed without merging**